internal/cli/cli.go

```diff
@@ -0,0 +1,112 @@
+package cli
+
+import (
+	"github.com/10gen/realm-cli/internal/terminal"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+// Command is an executable CLI command
+// This interface maps 1:1 to Cobra's Command.RunE phase
+type Command interface {
+	Handler(profile *Profile, ui terminal.UI, args []string) error
+}
+
+// CommandDefinition is a command's definition that the CommandFactory
+// can build a *cobra.Command from
+type CommandDefinition struct {
+	Command Command
+
+	// Description is the short description shown in the command's help text
+	// This value maps 1:1 to Cobra's `short` property
+	Description string
+
+	// Usage is the one-line usage message
+	// This value maps 1:1 to Cobra's `use` property
+	Usage string
+}
+
+// CommandFactory is a command factory
+type CommandFactory struct {
+	Profile *Profile
+	Config  Config
+}
+
+// Build builds a Cobra command from the specified CommandDefinition
+func (factory CommandFactory) Build(provider func() CommandDefinition) *cobra.Command {
+	cmdDef := provider()
+
+	cmd := cobra.Command{
+		Use:   cmdDef.Usage,
+		Short: cmdDef.Description,
+		RunE: func(c *cobra.Command, a []string) error {
+			return cmdDef.Command.Handler(
+				factory.Profile,
+				newCobraUI(factory.Config.UI, c),
+				a,
+			)
+		},
+	}
+
+	if preparer, ok := cmdDef.Command.(CommandPreparer); ok {
+		cmd.PreRunE = func(c *cobra.Command, a []string) error {
+			return preparer.Setup(
+				factory.Profile,
+				newCobraUI(factory.Config.UI, c),
+				factory.Config.Command,
+			)
+		}
+	}
+
+	if responder, ok := cmdDef.Command.(CommandResponder); ok {
+		cmd.PostRunE = func(c *cobra.Command, a []string) error {
+			err := responder.Feedback(
+				factory.Profile,
+				newCobraUI(factory.Config.UI, c),
+			)
+			if err != nil {
+				return err // TODO(REALMC-XXXX): handle this error gracefully
```

Comment:
real ticket? 😬

Reply:
there were a few of these... I'm going to go through now and create them.  Just held off until your review, to make sure none of these TODOs were actual blockers for this merge